File: java/ch17-01-tests/src/test/java/com/thesoftwaregorilla/tdd/money/SumTest.java
package com.thesoftwaregorilla.tdd.money;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.junit.jupiter.api.Assertions.*;

//<editor-fold desc="TO DO List">
//TODO: Look into refactoring the code in Sum to use BinaryOperator and/or UnaryOperator instead of Expression.
// (see issue #1)
//TODO: Refactor the testPlus and testTimes methods to use a single method with parameters for the operation type.
// (needs to wait for plus cleanup)
//</editor-fold>

@DisplayName("Sum Class Tests")
class SumTest {

    @DisplayName("reduce same currency")
    @ParameterizedTest(name = "{0} : {1}, {2} : {3}, Expected: {4}")
    @CsvSource({
            "USD, 5, USD, 5, 10",
            "CHF, 5, CHF, 5, 10",
            "ZAR, 5, ZAR, 5, 10"
    })
    void testReduceSameCurrency(String fromCurrency, int fromAmount, String toCurrency, int toAmount, int expectedAmount) {
        testReduce(fromCurrency, fromAmount, toCurrency, toAmount, expectedAmount);
    }

    @DisplayName("reduce different currency")
    @ParameterizedTest(name = "{0} : {1}, {2} : {3}, Expected: {4}")
    @CsvSource({
            "CHF, 10, USD, 5, 10",
            "ZAR, 20, CHF, 5, 6",
            "ZAR, 34, USD, 5, 7"
    })
    void testReduceDiffCurrency(String fromCurrency, int fromAmount, String toCurrency, int toAmount, int expectedAmount) {
        testReduce(fromCurrency, fromAmount, toCurrency, toAmount, expectedAmount);
    }

    private void testReduce(String fromCurrency, int fromAmount, String toCurrency, int toAmount, int expectedAmount) {
        Sum sum = new Sum(MoneyTest.getCurrencyFactory(fromCurrency).apply(fromAmount),
                MoneyTest.getCurrencyFactory(toCurrency).apply(toAmount));
        Money reduced = sum.reduce(BankTest.getBankWithRates(), toCurrency);
        assertEquals(MoneyTest.getCurrencyFactory(toCurrency).apply(expectedAmount), reduced);
    }

    @DisplayName("plus returning a sum (testPlusReturnsSum method in TDD book was part of MoneyTest)")
    @ParameterizedTest(name = "{0} : {1}")
    @CsvSource({
            "USD, 5",
            "CHF, 5",
            "ZAR, 5"
    })
    public void testPlusReturnsSum(String currency, int amount) {
        Money money = MoneyTest.getCurrencyFactory(currency).apply(amount);
        Expression result = money.plus(money);
        assertInstanceOf(Sum.class, result);
        Sum sum = (Sum) result;
        assertEquals(money, sum.augend);
        assertEquals(money, sum.addend);
    }


    @DisplayName("plus")
    @ParameterizedTest(name = "{0} : {1}")
    @CsvSource({
            "USD, 5",
            "CHF, 7",
            "ZAR, 10"
    })
    void testPlus(String currency, int amount) {
        final int expectedAmount = amount * 3;
        Expression money = MoneyTest.getCurrencyFactory(currency).apply(amount);
        Sum sum = new Sum(money, money);
        Expression result = sum.plus(money);
        assertInstanceOf(Sum.class, result);
        assertEquals(MoneyTest.getCurrencyFactory(currency).apply(expectedAmount), result.reduce(BankTest.getBankWithRates(), currency));
    }

    @DisplayName("times")
    @ParameterizedTest(name = "{0} : {1}, Multiplier: {2}")
    @CsvSource({
            "USD, 5, 3",
            "CHF, 7, 5",
            "ZAR, 10, 30"
    })
    void testTimes(String currency, int amount, int multiplier) {
        final int expectedAmount = (amount + amount) * multiplier;
        Expression money = MoneyTest.getCurrencyFactory(currency).apply(amount);
        Sum sum = new Sum(money, money);
        Expression result = sum.times(multiplier);
        assertInstanceOf(Sum.class, result);
        assertEquals(MoneyTest.getCurrencyFactory(currency).apply(expectedAmount), result.reduce(BankTest.getBankWithRates(), currency));
    }

}

File: cs/ch17-01-tests/Money.Tests/SumTest.cs
namespace TheSoftwareGorilla.TDD.Money.Tests;

#region TO DO's
#endregion

[TestFixture]
[Description("Sum Tests")]
public class SumTest
{
    [TestCase("USD", 5, "USD", 5, 10, TestName = "reduce same currency USD 5")]
    [TestCase("CHF", 5, "CHF", 5, 10, TestName = "reduce same currency USD 5")]
    [TestCase("ZAR", 5, "ZAR", 5, 10, TestName = "reduce same currency USD 5")]
    [TestCase("CHF", 10, "USD", 5, 10, TestName = "reduce different currency CHF 10 with USD 5, expected 10")]
    [TestCase("ZAR", 20, "CHF", 5, 6, TestName = "reduce different currency ZAR 20 with CHF 5, expected 6")]
    [TestCase("ZAR", 34, "USD", 5, 7, TestName = "reduce different currency ZAR 34 with USD 5, expected 7")]
    [Category("reduction")]
    public void TestReduce(string fromCurrency, int fromAmount, string toCurrency, int toAmount, int expected)
    {
        var sum = new Sum(MoneyTest.GetCurrencyFactory(fromCurrency).Invoke(fromAmount), MoneyTest.GetCurrencyFactory(toCurrency).Invoke(toAmount));
        var result = sum.Reduce(BankTest.GetBankWithRates(), toCurrency);
        Assert.That(result, Is.EqualTo(MoneyTest.GetCurrencyFactory(toCurrency).Invoke(expected)));
    }


    [TestCase("USD", 5, TestName = "plus returns sum USD 5")]
    [TestCase("CHF", 5, TestName = "plus returns sum CHF 5")]
    [TestCase("ZAR", 5, TestName = "plus returns sum ZAR 5")]
    [Category("arithmetic")]
    public void TestPlusReturnsSum(string currency, int amount)
    {
        var money = MoneyTest.GetCurrencyFactory(currency).Invoke(amount);
        var result = money.Plus(money);
        Assert.That(result, Is.InstanceOf<Sum>());
        Sum sum = (Sum) result;
        Assert.That(sum.Augend, Is.EqualTo(money));
        Assert.That(sum.Addend, Is.EqualTo(money));
    }

    [TestCase("USD", 5, TestName = "plus with new Sum USD 5")]
    [TestCase("CHF", 5, TestName = "plus with new Sum CHF 5")]
    [TestCase("ZAR", 5, TestName = "plus with new Sum ZAR 5")]
    [Category("arithmetic")]
    public void TestPlus(string currency, int amount)
    {
        int expected = amount * 3;
        var money = MoneyTest.GetCurrencyFactory(currency).Invoke(amount);
        var sum = new Sum(money, money);
        var result = sum.Plus(money);
        Assert.That(result, Is.InstanceOf<Sum>());
        Assert.That(result.Reduce(BankTest.GetBankWithRates(), currency), Is.EqualTo(MoneyTest.GetCurrencyFactory(currency).Invoke(expected)));
    }

    [TestCase("USD", 5, 3, TestName = "times with new Sum USD 5")]
    [TestCase("CHF", 7, 5, TestName = "times with new Sum CHF 7")]
    [TestCase("ZAR", 10, 30, TestName = "times with new Sum ZAR 10")]
    [Category("arithmetic")]
    public void TestTimes(string currency, int amount, int multiplier)
    {
        int expected = (amount + amount) * multiplier;
        var money = MoneyTest.GetCurrencyFactory(currency).Invoke(amount);
        var sum = new Sum(money, money);
        var result = sum.Times(multiplier);
        Assert.That(result, Is.InstanceOf<Sum>());
        Assert.That(result.Reduce(BankTest.GetBankWithRates(), currency), Is.EqualTo(MoneyTest.GetCurrencyFactory(currency).Invoke(expected)));
    }    
    
}